src/backings/tree.ts

```diff
@@ -0,0 +1,376 @@
+import {Proof, Tree} from "@chainsafe/persistent-merkle-tree";
+import {CompositeValue, ArrayLike, List} from "../interface";
+import {
+  BasicArrayType,
+  BasicListType,
+  CompositeArrayType,
+  CompositeListType,
+  CompositeType,
+  ContainerType,
+  isBasicType,
+  isCompositeType,
+  isContainerType,
+  isListType,
+  isVectorType,
+} from "../types";
+import {isTree} from "../util/tree";
+import {byteArrayEquals} from "../util/byteArray";
+import {Path} from "./backedValue";
+
+export function isTreeBacked<T extends CompositeValue>(value: unknown): value is ITreeBacked<T> {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return value && (value as any).type && (value as any).tree && isTree((value as any).tree);
+}
+
+export type KeyOf = string | number;
+export type ValueOf<T extends CompositeValue, V extends keyof T = keyof T> = ITreeBacked<T[V]> | T[V];
+
+export type TreeBacked<T extends CompositeValue> = ITreeBacked<T> & T;
+
+/**
+ * The ITreeBacked interface gives convenient access to a `CompositeType` + `Tree` as a wrapper object
+ *
+ * This is an alternative way of calling `tree_*` methods of the type
+ */
+export interface ITreeBacked<T extends CompositeValue> {
+  type: CompositeType<T>;
+  /**
+   * The merkle tree backing
+   */
+  tree: Tree;
+
+  /**
+   * Equality
+   *
+   * If both values are tree-backed, use equality by merkle root, else structural equality
+   */
+  equals(other: T): boolean;
```

Comment:
shouldn't other be of type ITreeBacked or unknown?

Reply:
I think type `T` makes sense here. `type TreeBacked<T> = ITreeBacked<T> & T`, so `TreeBacked` values will be included.